### ISSUE TITLE
[xray] Reusing cluster leads to hanging

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#2564]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: DL AMI
- **Ray installed from (source or binary)**: source
- **Ray version**: master
- **Python version**: 2.7
- **Exact command to reproduce**:

Distributed SGD code on ceddc26.

### Describe the problem
Ran the code once, program errored, then ran it again without restarting ray cluster, then this happened.

### Source code / logs
```
(tensorflow_p27) ubuntu@ip-172-31-56-124:/tmp/raylogs$ head -n 300 raylet_0-2018-08-04_05-38-59-07715.err
/home/ubuntu/ray-1/src/ray/object_manager/object_manager.cc:638: Create Chunk Failed index = 0: object already exists in the plasma store
WARNING:tensorflow:From /home/ubuntu/osdi2018/tfbench/convnet_builder.py:372: calling reduce_mean (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
+++ dirname /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/build.sh
++ cd /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow
++ pwd
+ PYARROW_TENSORFLOW_DIR=/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow
++ python -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.get_compile_flags()))'
+ TF_CFLAGS='-I/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include -D_GLIBCXX_USE_CXX11_ABI=0'
++ python -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.get_link_flags()))'
+ TF_LFLAGS='-L/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow -ltensorflow_framework'
++ uname
+ '[' Linux == Darwin ']'
+ NDEBUG=-DNDEBUG
+ g++ -std=c++11 -g -shared /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc -o /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.so -DNDEBUG -I/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include -L/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow -lplasma -larrow_python -larrow -fPIC -I/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include -D_GLIBCXX_USE_CXX11_ABI=0 -L/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow -ltensorflow_framework -DGOOGLE_CUDA -O2
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:65:0: warning: "DCHECK" redefined
 #define DCHECK(condition)      \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:273:0: note: this is the location of the previous definition
 #define DCHECK(condition) \
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:71:0: warning: "DCHECK_EQ" redefined
 #define DCHECK_EQ(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:283:0: note: this is the location of the previous definition
 #define DCHECK_EQ(x, y) _TF_DCHECK_NOP(x, y)
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:74:0: warning: "DCHECK_NE" redefined
 #define DCHECK_NE(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:284:0: note: this is the location of the previous definition
 #define DCHECK_NE(x, y) _TF_DCHECK_NOP(x, y)
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:77:0: warning: "DCHECK_LE" redefined
 #define DCHECK_LE(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:285:0: note: this is the location of the previous definition
 #define DCHECK_LE(x, y) _TF_DCHECK_NOP(x, y)
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:80:0: warning: "DCHECK_LT" redefined
 #define DCHECK_LT(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:286:0: note: this is the location of the previous definition
 #define DCHECK_LT(x, y) _TF_DCHECK_NOP(x, y)
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:83:0: warning: "DCHECK_GE" redefined
 #define DCHECK_GE(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:287:0: note: this is the location of the previous definition
 #define DCHECK_GE(x, y) _TF_DCHECK_NOP(x, y)
 ^
In file included from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/common.h:32:0,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/plasma/client.h:30,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:38:
/home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/include/arrow/util/logging.h:86:0: warning: "DCHECK_GT" redefined
 #define DCHECK_GT(val1, val2) \
 ^
In file included from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/logging.h:25:0,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/lib/core/refcount.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/tensor_coding.h:21,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/resource_handle.h:19,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/allocator.h:24,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/tensor.h:20,
                 from /home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/framework/device_base.h:23,
                 from /home/ubuntu/ray-1/python/ray/pyarrow_files/pyarrow/tensorflow/plasma_op.cc:18:
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/tensorflow/include/tensorflow/core/platform/default/logging.h:288:0: note: this is the location of the previous definition
 #define DCHECK_GT(x, y) _TF_DCHECK_NOP(x, y)
 ^
/home/ubuntu/ray-1/src/ray/object_manager/object_manager.cc:638: Create Chunk Failed index = 0: object already exists in the plasma store
/home/ubuntu/ray-1/src/ray/object_manager/object_manager.cc:638: Create Chunk Failed index = 0: object already exists in the plasma store
/home/ubuntu/ray-1/src/ray/object_manager/object_manager.cc:638: Create Chunk Failed index = 0: object already exists in the plasma store
/home/ubuntu/ray-1/src/ray/object_manager/object_manager.cc:638: Create Chunk Failed index = 0: object already exists in the plasma store
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/thirdparty/build/arrow/cpp/src/plasma/client.cc:960 Check failed: manager_conn_ >= 0
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
/home/ubuntu/ray-1/src/ray/raylet/node_manager.cc:1112: Task re-execution is not currently implemented
...
```

**Comment by elibol** [comment]
@pcmoritz I recall seeing the check failed error whenever `fetch` was called directly with `xray`; this was when `fetch` hadn't yet been implemented. Does that check fail when a plasma client method that depends on the original object manager is invoked?

The node manager message appears to be triggered due to a call to `ResubmitTask`, which is unimplemented. The call to `ResubmitTask` appears to stem from a sequence of calls beginning with a notification from the task lease table: https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L160-L162

@stephanie-wang can you provide some thoughts on what triggers task re-submission?

**Comment by stephanie-wang** [comment]
@richardliaw are all of the raylets still alive? Task resubmission might get triggered if there is no raylet executing a task that's required by a live raylet (e.g., because the raylet that was supposed to execute the task died).

**Comment by richardliaw** [comment]
OK After talking to @stephanie-wang seems like this main issue is just that the Driver finishing does not clean up the rest of the resources/actors associated with the actor.

**Comment by robertnishihara** [comment]
When this issue is addressed, we should enable the test in #2797. cc @zhijunfu

**Comment by robertnishihara** [comment]
@zhijunfu I created https://github.com/ray-project/ray/issues/2809 to keep track of what needs to be cleaned up when a driver exits.

**Comment by zhijunfu** [comment]
@robertnishihara  Thanks for the information, Robert.

**Comment by robertnishihara** [comment]
Reusing a cluster has been fixed by #2826. @richardliaw can you confirm?
